File: app/src/main/java/me/shreyasr/chatse/chat/ChatWebSocketListener.java
package me.shreyasr.chatse.chat;

import android.util.Log;

import com.squareup.okhttp.Response;
import com.squareup.okhttp.ws.WebSocket;
import com.squareup.okhttp.ws.WebSocketListener;

import org.codehaus.jackson.JsonNode;
import org.codehaus.jackson.map.ObjectMapper;

import java.io.IOException;

import okio.Buffer;
import okio.BufferedSource;

public class ChatWebSocketListener implements WebSocketListener {

    private ChatActivityFragment fragment;
    private final ObjectMapper mapper;
    private final int roomNum;

    public ChatWebSocketListener(ChatActivityFragment fragment, ObjectMapper mapper, int roomNum) {
        this.fragment = fragment;
        this.mapper = mapper;
        this.roomNum = roomNum;
    }

    @Override public void onOpen(WebSocket webSocket, Response response) {
        Log.e("ws", "ws open");
    }

    @Override public void onFailure(IOException e, Response response) {
        Log.e("ws", "ws fail");
        Log.e(e.getClass().getSimpleName(), e.getMessage(), e);
    }

    @Override public void onMessage(BufferedSource payload,
                                    WebSocket.PayloadType type) throws IOException {
        Log.e("ws", "ws message");
        String message = payload.readUtf8();
        payload.close();
        Log.e("ws", "recv: " + message);
        try {
            JsonNode root = mapper.readTree(message);
            if (!root.has("r" + roomNum)) {
                Log.e("No room element", root.toString());
                return;
            }
            JsonNode roomNode = root.get("r" + roomNum);
            if (!roomNode.has("e")) return;

            fragment.handleNewEvents(roomNode.get("e"));
        } catch (IOException e) {
            Log.e(e.getClass().getSimpleName(), e.getMessage(), e);
        }
    }

    @Override public void onPong(Buffer payload) {
        Log.e("ws", "ws pong");
    }

    @Override public void onClose(int code, String reason) {
        Log.e("ws", "ws close");
    }
}

File: app/src/main/java/com/tristanwiley/chatse/chat/service/ChatWebSocketListener.kt
package com.tristanwiley.chatse.chat.service

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.squareup.okhttp.Response
import com.squareup.okhttp.ws.WebSocket
import com.squareup.okhttp.ws.WebSocketListener
import okio.Buffer
import okio.BufferedSource
import timber.log.Timber

import java.io.IOException

/**
 * WebSocketListener that listens to websocket
 */
class ChatWebSocketListener(private val site: String, private val listener: ChatWebSocketListener.ServiceWebsocketListener) : WebSocketListener {
    private val mapper = ObjectMapper()

    override fun onOpen(webSocket: WebSocket, response: Response) {
        Timber.i("ChatWebSocketListener websocket open: $site")
        listener.onConnect(site, true)
    }

    override fun onFailure(e: IOException, response: Response?) {
        listener.onConnect(site, false)
    }

    @Throws(IOException::class)
    override fun onMessage(payload: BufferedSource,
                           type: WebSocket.PayloadType) {
        val message = payload.readUtf8()
        payload.close()
        try {
            val root = mapper.readTree(message)
            listener.onNewEvents(site, root)
        } catch (e: IOException) {
            Timber.e("ChatWebSocketListener${e.message}")
        }

    }

    override fun onPong(payload: Buffer) {
        Timber.i("ChatWebSocketListener web socket pong: $site")
    }

    override fun onClose(code: Int, reason: String) {
        Timber.i("ChatWebSocketListenerweb socket close: $site: $code, $reason")
    }

    interface ServiceWebsocketListener {
        fun onNewEvents(site: String, root: JsonNode)

        fun onConnect(site: String, success: Boolean)
    }
}
